### PR TITLE
Implement plain `parallel:` keyword

### LINKED ISSUE #1227: $CI_NODE_INDEX defaults to the string "null" when not in a parallel job, expected the empty string
**Minimal .gitlab-ci.yml illustrating the issue**
```yml
---
example:
  script:
    - echo $CI_NODE_INDEX
    - echo $CI_NODE_TOTAL
    - echo $This_variable_does_not_exist
```

**Expected behavior**
Snippet of gitlab.com job output:
```
$ echo $CI_NODE_INDEX
$ echo $CI_NODE_TOTAL
1
$ echo $This_variable_does_not_exist
Cleaning up project directory and file based variables
00:01
Job succeeded
```

**Host information**
Ubuntu 20.04
gitlab-ci-local 4.49.0

**Containerd binary**
Are you using docker or podman, or any other relevant containerization tool. - No

**Additional context**
Looks like the specific lines that set the variable are here:

https://github.com/firecow/gitlab-ci-local/blob/d6378db7dc6975c92356d16be0d7b9f563f7d9d9/src/job.ts#L29

https://github.com/firecow/gitlab-ci-local/blob/d6378db7dc6975c92356d16be0d7b9f563f7d9d9/src/job.ts#L172

### PR DESCRIPTION
**Pull request by PigeonF**
See <https://docs.gitlab.com/ee/ci/yaml/#parallel>.

When writing the test I saw that there was some verification of the schema (e.g. I could not set `parallel: 1`), but when running `gitlab-ci-local` directly it works fine. Is this intended behaviour? There is now an edge case here, which I cannot test for at the moment: Setting `parallel: 1` works on <https://gitlab.com>, but I cannot write a test for it because it does not match the official schema (which is apparently wrong, I have filed <https://gitlab.com/gitlab-org/gitlab/-/issues/462979>).

This fixes #1227 .

### CODE REVIEW

File: src/job.ts [hunk 1]
```diff
@@ -169,7 +169,9 @@ export class Job {
         predefinedVariables["CI_ENVIRONMENT_NAME"] = this.environment?.name ?? "";
         predefinedVariables["CI_ENVIRONMENT_SLUG"] = this.environment?.name?.replace(/[^a-z\d]+/ig, "-").replace(/^-/, "").slice(0, 23).replace(/-$/, "").toLowerCase() ?? "";
         predefinedVariables["CI_ENVIRONMENT_URL"] = this.environment?.url ?? "";
-        predefinedVariables["CI_NODE_INDEX"] = `${opt.nodeIndex}`;
+        if (opt.nodeIndex !== null) {
+            predefinedVariables["CI_NODE_INDEX"] = `${opt.nodeIndex}`;
+        }
```
Comment (PigeonF) [inline src/job.ts:174]: I tested on gitlab.com, and the variable is not set (i.e. it is not present in the environment, not set to an empty string).

Looking at the code around this though, it seems the other predefined variables are set to an empty string instead. I don't care either way, but I just wanted to point this out (since it breaks the surrounding code style).

**Comment by ANGkeith** [comment]
> When writing the test I saw that there was some verification of the schema (e.g. I could not set parallel: 1), but when running gitlab-ci-local directly it works fine. Is this intended behaviour? 

probably not.

The json schema validation implemented in https://github.com/firecow/gitlab-ci-local/pull/1167 should be able to catch this.
For now, it's still hidden behind the option, `export GCL_ENABLE_JSON_SCHEMA_VALIDATION=true`

File: src/job.ts (hunk 1 above)
Comment (ANGkeith) [reply src/job.ts:174]: seems fine to me. 

But for regression testing, could you add a simple test case for this bug fix ?

File: src/parallel.ts [hunk 2]
```diff
@@ -1,6 +1,28 @@
...
+        }
+    }
+
+    return {
+        matrixVariables: [null],
+        nodesTotal: null,
+    };
+}
+
 export function matrixVariablesList (jobData: any, jobName: string) {
```
Comment (ANGkeith) [inline src/parallel.ts:26]:
```suggestion
export function matrixVariablesList (jobData: any, jobName: string) {
      if (Number.isInteger(jobData.parallel)) {
          return Array(jobData.parallel).fill(null);
      }
```

Comment (ANGkeith) [reply src/parallel.ts:26]: let's simplify the logic like that

**Comment by ANGkeith** [comment]
on a side note, @PigeonF  i think you're a collaborator right?  

do you mind pushing directly into `github.com/firecow` ?

File: tests/test-cases/parallel/integration.parallel.test.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import {WriteStreamsMock} from "../../../src/write-streams";
+import {handler} from "../../../src/handler";
+import chalk from "chalk";
+
+test("parallel <test-job>", async () => {
+    const writeStreams = new WriteStreamsMock();
+    await handler({
+        cwd: "tests/test-cases/parallel",
+        jobs: ["test-job"],
```
Comment (ANGkeith) [inline tests/test-cases/parallel/integration.parallel.test.ts:9]:
```suggestion
        job: ["test-job"],
```

**Comment by ANGkeith** [comment]
> Setting parallel: 1 works on https://gitlab.com/, but I cannot write a test for it because it does not match the official schema

we've copied the json schema  instead of pulling from gitlab.com (maybe we should ?),

but for now i think you can fix the schema @ https://github.com/firecow/gitlab-ci-local/blob/master/src/schema/schema.json#L1275

and your test should be good to go

**Comment by PigeonF** [comment]
> on a side note, @PigeonF i think you're a collaborator right?
> 
> do you mind pushing directly into `github.com/firecow` ?

Is there some way to change the source branch of the PR directly? Or do I close this one and create a new one from the new branch?

**Comment by ANGkeith** [comment]
> > on a side note, @PigeonF i think you're a collaborator right?
> > do you mind pushing directly into `github.com/firecow` ?
> 
> Is there some way to change the source branch of the PR directly? Or do I close this one and create a new one from the new branch?

for this PR it's fine, but for future PR, will be nice if you push directly to `github.com/firecow`

File: src/parallel.ts [hunk 4]
```diff
@@ -1,29 +1,10 @@
...
-    return {
-        matrixVariables: [null],
-        nodesTotal: null,
-    };
-}
-
 export function matrixVariablesList (jobData: any, jobName: string) {
+    if (Number.isInteger(jobData.parallel)) {
+        return Array(jobData.parallel).fill(null);
+    }
```
Comment (PigeonF) [inline src/parallel.ts:7]: I have implemented this simplified way like you suggested. Tell me which version you prefer and I will either drop this commit or squash it with the previous one.

Comment (ANGkeith) [reply src/parallel.ts:7]: personally, i'll prefer this

**Review by ANGkeith — APPROVED** [review]

**Comment by ANGkeith** [comment]
final small refactoring https://github.com/PigeonF/gitlab-ci-local/pull/1

**Review by ANGkeith — APPROVED** [review]
